Add support for ExcessMapEstimator on HLI

registerrier opened the pull request:
<!-- These are hidden commments. You can leave them or delete them. -->

<!-- Don't be shy! Github issues or pull requests are welcome any time. -->
<!-- But if you first want to chat, use the Gammapy Slack. -->
<!-- See https://gammapy.org/contact.html -->

**Description**
<!-- What is the motivation for this pull request? -->
<!-- Briefly: what changes are done here? -->
<!-- If this is to address a Github issue, metion it's number to create a link -->
<!-- Example: This PR is to fix issue #42 -->

This pull request introduces a new config argument for `ExcessMapEstimator` as well as a the corresponding method `Analysis.get_excess_map()`

The default values for `EnergyAxisConfig` arguments is set to None, but default values are added to `EnergyAxesConfig`
**Dear reviewer**
<!-- Let the reviewer and Gammapy team know what you want: -->
<!-- * Is this ready for review? Or is it work in progress and you want some feedback? -->
<!-- * Do you want ot go through review here? Or if someone just finish this up and merge it in? -->
<!-- Do you have any specific questions, e.g. about API or implementation? -->
<!-- Do you include a test executing new code you're adding (to make sure it runs)? -->
<!-- Do you include some documentation? Is it needed? -->

Bultako commented on gammapy/analysis/config.py:
```diff
@@ -138,6 +138,12 @@ class FitConfig(GammapyBaseConfig):
 
+class ExcessMapConfig(GammapyBaseConfig):
+    correlation_radius: AngleType = '0.1 deg'
+    parameters: dict = {"selection_optional": ""}
```
I think `"selection_optional": ""` can be safely removed, there are other possible parameters allowed, so better declare only those that we want to set by default. Let's make `parameters: dict = {}` like in the other classes.

registerrier replied:
Indeed. The default is now `selection_optional=None` so we don't need to pass anything.

Bultako commented on gammapy/analysis/config.py:
```diff
@@ -138,6 +138,12 @@ class FitConfig(GammapyBaseConfig):
 
 
+class ExcessMapConfig(GammapyBaseConfig):
+    correlation_radius: AngleType = '0.1 deg'
```
"0.1 deg" double quotes

registerrier replied:
Done

Bultako reviewed (changes requested):
Thanks @registerrier I've added some comments in-between the lines of code.

registerrier commented on gammapy/analysis/core.py:
```diff
@@ -384,12 +407,15 @@ def _spectrum_extraction(self):
-            interp="log",
-            node_type="edges",
-        )
+        if axis.min is None and axis.max is None:
```
I wanted to have the default energy axis to be `None` so that we could pass None arguments to estimators.  Hence, raising `ValueError` is not completely adapted.
I followed your second suggestion. 
We might add a `log.info` or `log.warning` when the axis is left to `None` since the type validation is not complete in this case?

Bultako replied:
Ok, thinking more carefully about this I guess we should better do the values checking validation directly in the `EnergyAxisConfig` class in the `config.py` file.  i.e. `min` value declared and `max` value missing, `nbins` missing when both `min` and `max` are declared, etc. for the moment we could just place a simple `log.info` here when we return `None`. Something as what is shown below.

```
if axis.min is None:
    log.info("No energy axis defined.") 
    return None
```

I will make another PR addressing complex validation among fields in the different classes of the model of the HLI, this concerns also other classes than this `EnergyAxisConfig`.

Bultako commented:
@registerrier 

Two more things I've found:
- I think it would be nice to add the new config block `excess_map` to the file `analysis/config/example-1d.yaml` and analysis/config/docs.yaml`, even if it s not needed, but I think it is good to have it (with the default values is ok) so we can easily inspect allowed params in a sample config file.. so we could maybe use it for doc purposes in the future.
- I see you use `config.excess_map`, `analysis.get_excess_map()`,  `analysis.excess_maps`, etc. could we be more consistent and use only the plural or singular form?

adonath commented on gammapy/analysis/core.py:
```diff
@@ -207,6 +207,29 @@ def get_flux_points(self):
+            energy_edges=energy_edges,
+            **excess_settings.parameters
+        )
+        self.excess_maps = excess_map_estimator.run(self.datasets[0])
```
I just asked myself which error is thrown if the `ExcessMapEstimator` is called with a `SpectrumDataset`. I found https://github.com/gammapy/gammapy/blob/master/gammapy/estimators/excess_map.py#L154, which is basically fine, but one could maybe improve the message a bit and say which dataset type was passed and which one is expected, just because the HLI is probably more likely to be used by beginners.

registerrier replied:
Indeed I changed the message to explicitly state that 1D datasets are not supported.

adonath reviewed (dismissed):
Thanks @registerrier , I have left on minor inline comment, otherwise looks fine...

registerrier commented:
> @registerrier
> 
> Two more things I've found:
> 
> * I think it would be nice to add the new config block `excess_map` to the files `analysis/config/example-1d.yaml` and `analysis/config/docs.yaml`, even if it s not needed, but I think it is good to have it (with the default values is ok) so we can easily inspect allowed params in a sample config file.. and maybe use it for doc purposes in the future.
> * I see you use `config.excess_map`, `analysis.get_excess_map()`,  `analysis.excess_maps`, etc. could we be more consistent and use only the plural or singular form?

Good points. I have updated `analysis/config/example-3d.yaml` and  `analysis/config/docs.yaml`.
I also changed the result product property to `Analysis.excess_map` for consistency.

adonath reviewed (approved):
Thanks @registerrier, I have no further comments.
